src/posts/react-native-monorepo.md

```diff
@@ -0,0 +1,180 @@
+three repositories to work with, it was too easy to forget to add one or the
+other (as you're generally developing with just one simulator open), and then
+it breaks the build for other environment.
+
+We'd also get into a state where the `master` branch of `react-native`
+contained some changes that had only been coordinated with e.g. the `ios`
+repo, and the `android` repo's `react-native` submodule would be several
+commits behind. Then someone working on the android side would update the
+submodule, and they'd have to track down all the breakages.
+
+In short, we started running into a lot of synchronization issues, and we
+really wanted to just have things be in the same place, such that pull
+requests could be combined, reviews would be more coherent, and it would be
+easier to verify correctness between codebases.
+
+### Anticipated Pros
+
+- changing js↔native apis would be easier because you only need a single
+  pull request instead of 3
+- as a result, they would be less likely to get out of sync
+- not having to mess with submodules 🎉
+
+### Anticipated Cons
+
+- we might lose git history (this didn't turn out to be the case)
+- we'd have to change all of our jenkins build scripts
+- moving all our developers to a new repository requires coordination
+- we'd lose any in-flight branches & open PRs to the old repositories (we
+  actually found a solution for this too)
+
+## What were the steps?
+
+### Setup
+
+Make a fresh monorepo
+
+```
+mkdir mobile; cd mobile; git init .
+```
+
+Have all the repos that you want to combine cloned & fully up to date
+
+```
+$ ls .
+mobile
+android
+ios
+react-native
+```
+
+### Preparing the repos for merging
+
+Clone each repo into `m_reponame` (using `android` as the example) and then
+move all files into a subfolder (except for `.git`, of course).
+
+```
+cd m_android
+mkdir android
+mv * .* android
+mv android/.git .
+```
+
+commit result; `git add . -m'move to subfolder'`
+
+With the code for each respective codebase moved into a subdirectory, we're
+then able to move them all into a single repository without having them clash
+with each other. To illustrate, here's what the rough directory structure
+looks like:
+
+```
+android/
+  build.gradle, etc.
+ios/
+  AppDelegate.m, etc.
+react-native/
+  index.ios.js, etc.
+m_android/
+  android/
+    build.gradle, etc.
+m_ios/
+  ios/
+    AppDelegate.m, etc.
+m_react-native/
+  react-native/
+    index.ios.js, etc.
+```
+
+The monorepo will then have the following structure:
+
+```
+mobile/
+  android/
+    build.gradle, etc.
+  ios/
+    AppDelegate.m, etc.
+  react-native/
+    index.ios.js, etc.
+```
+
+### Merge each `m_repo` into the monorepo
+
+Turns out git has super powers, and can totally merge in multiple unrelated
+repositories and preserve all the relevant git history. Who knew?
+
+```
+cd mobile
+git fetch ../m_android
+g merge FETCH_HEAD --no-ff --allow-unrelated-histories -m 'merging in android repo'
```

Comment:
What's `g`?  Did you mean `git` or is it something else?

Also, such a long command will result in scrolling when we actually publish this.  I think you'll want to do
```
g merge FETCH_HEAD --no-ff --allow-unrelated-histories \
    -m 'merging in android repo'
```
-- that is, make it two lines long.